### ISSUE TITLE
Select text from PDF

### ISSUE THREAD

**Issue by sotte** [xournalpp/xournalpp#1745]
**Is your feature request related to a problem? Please describe.**
xournalpp is great for annotating PDFs (highligthing, adding text, whatnot) but often I want to **select text from the PDF, copy it,** and paste it into another program. xournalpp doesn't have a text selection tool like most PDF readers.

**Describe the solution you'd like**
Add a text selection tool.

**Describe alternatives you've considered**
Currently I open a PDF reader to select the text that way.

**Comment by easyw** [comment]
@sotte 
this feature is the most one I'm missing too...
+1 from my side

**Comment by caliarim** [comment]
+1

**Comment by DystopianLphant** [comment]
+1, I have to keep my pdf viewer opened to copy the text I need

**Comment by sotte** [comment]
Any ideas how to proceed here? "Simple text selection" can be quite tricky (multi column documents). Could we leverage the underlying pdf engine?

**Comment by LittleHuba** [comment]
You can use the underlying pdf engine theoretically. In the current code of Xournal++ this is hard though, as the PDF is rendered as background and after that all state of the engine is discarded.

**Comment by ozel** [comment]
same here. the only occasion that I am using another PDF reader sometimes is for coping things like title or author names for completing the Zotero database entry. Since the text search feature works pretty well, it looks like not all of the PDF engines state is discarded... any pointers on where to start looking in the source code?

**Comment by ReddingZH** [comment]
This feature would be a huge plus for xournalpp and its use cases.

**Comment by zeidoo** [comment]
+1. Would be really nice.

**Comment by All3xJ** [comment]
up. It should be a must

**Comment by Rots** [comment]
As per recommendation of @rolandlo just a notification in this thread that there are also annotation use cases with PDFs. I've written down my use case [in discussions](https://github.com/xournalpp/xournalpp/discussions/2681).

**Comment by rmarabini** [comment]
+1

**Comment by AndersSpringborg** [comment]
+1

**Comment by VivekThazhathattil** [comment]
In the lightweight pdf reader MuPDF, text selection is implemented as simply right clicking and dragging to  draw a rectangle over the text that you want to copy and then pressing ctrl+C to copy it. Taking inspiration from this, I tried tweaking xournalpp's code to allow select rectangle tool to act in a similar way. Then I was able to use poppler api command poppler_page_get_selected_text() given the popplerRectangle to get  the required text. Maybe a similar method would do the trick!

**Comment by adhuliya** [comment]
+1
As a quick fix, can we add a shortcut to open the current location of the pdf in a lightweight pdf viewer to copy the text?
This will save the pain of,
1. going to the file
2. opening it with a pdf viewer
3. going to the location of the text
4. copy the text!

The recommended shortcut can eliminate steps 1 to 3.
A lightweight pdf viewer can be made a part of the package itself for this purpose.

Is this functionality already there?

**Comment by Technius** [comment]
I think it would be possible to implement a plugin with that functionality.

**Comment by adhuliya** [comment]
I took this time to study the plugin system in xournalpp and it seems this is feasible. Adding a quick solution might be possible, but adding a reliable cross platform support seems non-trivial (which pdf reader to invoke, what arguments does it take etc. needs to be captured, catalogued and tested). Moreover, invoking heavy pdf readers (for example foxit) just for this purpose seems impractical (as some user systems might only have these heavy pdf readers).

On the other hand, if some internal sub-system of xournalpp could be used to render the few paragraph that the user
has on the screen (say in a separate popup), then that could be very reliable and accurate across platforms.
This can probably be possible with a plugin as well, if the required APIs are available. Moreover, instead of a popup,
temporarily disabling the layering could expose the PDF for text selection (I am not sure if this is possible,
but I am loud thinking here of a possibility of bringing the PDF from the background to foreground.)

xournalpp has become a an indispensable part of my work pipeline. Hence, I would like to contribute to this
issue. If someone is already developing a solution and needs some external help, I can be available from
this 10th of July for co-development. If no concrete solution is in the pipeline, I can start exploring the same
from next month.

**Comment by aeghn** [comment]
I also recently started using Xournalpp and started trying to solve this problem。


https://user-images.githubusercontent.com/51843252/128286156-55f5985e-3f16-41fc-bf93-1885b3fcdded.mp4

**Comment by aeghn** [comment]
Hey, guys, I think I have almost finished this, here is the preview.

https://user-images.githubusercontent.com/51843252/128634365-ddc3e7fd-255a-4ea9-933e-4a1c1257afaf.mp4

This is my first experience with c++ programs and also with gtk programs (in fact I have no previous experience with gui programs at all), so there may be some problems with it, and based on that I decided to open a pr after some more personal use.

link: https://github.com/aeghn/xournalpp/tree/dev_20210802_select_add_element

So any help, tests and suggestions are appreciated!

**Comment by Technius** [comment]
@aeghn This looks fantastic! I'm looking forward to seeing your PR. A few quick questions:

* Are the popup menu actions creating new elements?
* How do you feel about having a separate tool for PDF interaction? Do you feel like it would be more convenient to use a gesture/modified click instead?

**Comment by aeghn** [comment]
@Technius thanks for your reply.
   
   
  

>  Are the popup menu actions creating new elements?

When I started trying to make this tool, I used GTKMenu and created a new menu every time, but then I found that using GTKMenu didn't harmonize with the whole xournalpp. So I looked at the floatingToolbox code and made a simple copy of it using GTKGrid.

So every time the menu pops up, it is displayed via `gtk_widget_show` to show the GTKGrid that was created when the app was launched.

> How do you feel about having a separate tool for PDF interaction? Do you feel like it would be more convenient to use a gesture/modified click instead?

I'm a little confused about _a gesture/modified click_. Could you please explain a little more?
In my opinion, an appropriate way to interact with pdf text is:

1.  select a region;
2.  check whether the region contains text,  if true:
3.  show the range of text selection;
4.  pop-up menu, the user selects what needs to do.

**Comment by Technius** [comment]
@aeghn 

Sorry, I think I didn't express my questions very clearly. So let me rephrase them a little:

* Are the popup menu actions creating new elements/objects to place on to the page? For example, does the strikethrough action create a stroke that is under the selected text?
* In addition to/as an alternative to a separate tool for PDF interaction, do you feel like it would be convenient to add a mouse/gesture modifier to open the PDF popup menu? For example, some users may find it convenient to open the popup menu with Alt+Click or by triple tapping. Or they may just want to use the existing `Select Object` tool with Ctrl/Alt/Shift held down to interact with the PDF. I'm not asking as a feature request; I just want your thoughts on this since you have a working prototype. There was a similar discussion in #2660 (which I did some initial work on but didn't have time to finish).

**Comment by aeghn** [comment]
@Technius 
1. Yes, the highlight/strikethrough/underline operations are all created stroke objects, I was thinking that these three operations are to help improve the efficiency of user operations, obviously selecting these texts and marking them with one click is a bit faster and more accurate than scratching around on the screen with the mouse, and logically I prefer to put these operations on the same level as manual marking, they do the the same, but the process is different. And using stroke objects directly would also reduce the code implementation.
2. Yes, I agree with you that it is good to have some key assistance, but on my current experience, with the `FloatingToolbox`, right-click to switch between different tools so that people feel the whole process and smooth and comfortable, and the use of shortcut keys experience has been similar. I am focusing PDF text selection now, which is necessarily accompanied by a text selection, and it is simliar to `RectSelection`/`RegionSelection`, just selecting different content, so I think these three operations should have the same entry. Based on this idea, I put this in the `Selection` class below, and with a new implementation of the `PdfTextSelectControl` to do the specific logic.

**Comment by leo-wiz** [comment]
@aeghn  Hello. Am I right, your copy/paste  tool hasn't got to main fork ? And your fork hasn't windows installer ? 
Thanks for your work.

**Comment by aeghn** [comment]
> @aeghn  Hello. Am I right, your copy/paste  tool hasn't got to main fork ? I can't find select tool as plugin either.  And your fork hasn't windows installer ? 

It looks like a maintainer is merging that pr.

So there will be a windows installer in the near future I guess.

I am preparing to build my own pdf reader tool based on pdf.js and will consider supporting more

1. support for the previously provided notes feature
2. introduce bookmarks, easy to search for tags and more
3. introduce features such as roam to associate different books and bookmarks
4. add a split-screen feature to read and take notes at the same time (probably org-mode)

**Comment by Technius** [comment]
This feature has been merged and will be available starting from tomorrow's nightly build. Big thanks to @aeghn for making it possible.

Usage:

* Two new tools have been introduced, "Select Linear PDF Text" and "Select PDF Text In Rectangle". The tools are only functional on documents with background PDFs. Linear selection allows users to select all text between a start and end point, while rectangle selection ("area selection") allows users to select all text that are contained within a selection rectangle.
* Using the PDF text tool, a user can use the primary input button to select some text on the background PDF file, similar to well-known PDF viewers. When the selection is "finalized" (e.g., by the user releasing the primary input button), a toolbox will be shown.
* The toolbox has buttons that allow a user to: 1) to copy the selected text, 2) generate a highlight stroke over the selected text; 3) generate a stroke to strike-through the text; and 4) generate a stroke to underline the text; and 5) toggle between linear and rectangle selection.
* By double pressing or triple pressing with the linear selection tool, a user may select by words or lines, respectively, instead of by character/symbol.

**Comment by BernardoFBBraga** [comment]
Hi, I just installed the nightly build following the instructions on the main page, but I can't seem to find the "select linear PDF text" tool. Where is it? 
Looking in the "about" section, seems like the version is outdated.
![image](https://user-images.githubusercontent.com/3694228/153213370-dcfe75a1-5f75-41c5-bf1a-8cd7705a0ce1.png)

I followed these instructions
```
sudo add-apt-repository ppa:andreasbutti/xournalpp-master
sudo apt update
sudo apt install xournalpp
```

Is there a different method of installing?

**Comment by rolandlo** [comment]
@BernardoFBBraga What OS do you use? If it is Ubuntu 18.04 (or derivative) then you won't have an up to date version via PPA (since the build is broken). You can use the [nightly release](https://github.com/xournalpp/xournalpp/releases/tag/nightly) instead (or the flatpak from flathub or snap package from snapcraft).

**Comment by BernardoFBBraga** [comment]
@rolandlo I found that nightly relase page too, but it says it was last released on July 2021, so I guess it's outdated.
I'm currently trying to build it myself but ran into this error:
```
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.12...3.10.2 or higher is required.  You are running version 3.10.2


-- Configuring incomplete, errors occurred!
```
So I'm trying to figure out how to update my cmake now. Seems weird that the error message tells me minimum version is the version I current have but it still doesn't work.

I'm using Linux Mint 19.3 Cinnamon

**Comment by BernardoFBBraga** [comment]
Oh maybe it's because the ubuntu build is broken ?

**Comment by rolandlo** [comment]
@BernardoFBBraga The nightly release assets are not outdated. The release date is wrong. It is not updated, just the assets are updated.
The PPA Ubuntu 18.04 build is broken in contrast and receives no updates.

About compiling from source: `cmake 3.12` is required (in minimum). I don't know why the error message says differently.

**Comment by Technius** [comment]
There was a recent change to `CMakeLists.txt`, but the minimum version line is wrong: https://github.com/xournalpp/xournalpp/blob/ebaed95f4784a4a5b6f276e6d3a9c3b339e60f31/CMakeLists.txt#L1

The minimum version line should be ` cmake_minimum_required(VERSION 3.12)`.

**Comment by Febbe** [comment]
> There was a recent change to `CMakeLists.txt`, but the minimum version line is wrong: https://github.com/xournalpp/xournalpp/blob/ebaed95f4784a4a5b6f276e6d3a9c3b339e60f31/CMakeLists.txt#L1
> 
> The minimum version line should be ` cmake_minimum_required(VERSION 3.12)`.

It is not wrong, its a new syntax to specify a range of supported cmake versions.

**Comment by Technius** [comment]
It is an incorrect use of the new syntax and should not have been used in the first place. The minimum version is nonsensical because the upper bound is set to the user's CMake version. If the user has CMake older than 3.12, then the lower bound is _greater_ than the upper bound. Hence the _extremely misleading_ error message observed:
```
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.12...3.10.2 or higher is required.  You are running version 3.10.2
```
It is sufficient to use `cmake_minimum_required(VERSION 3.12)`.

**Comment by thierrybo** [comment]
> This feature has been merged and will be available starting from tomorrow's nightly build. Big thanks to @aeghn for making it possible.
> 
> Usage:
> 
>     * Two new tools have been introduced, "Select Linear PDF Text" and "Select PDF Text In Rectangle". The tools are only functional on documents with background PDFs. Linear selection allows users to select all text between a start and end point, while rectangle selection ("area selection") allows users to select all text that are contained within a selection rectangle.
> 
>     * Using the PDF text tool, a user can use the primary input button to select some text on the background PDF file, similar to well-known PDF viewers. When the selection is "finalized" (e.g., by the user releasing the primary input button), a toolbox will be shown.
> 
>     * The toolbox has buttons that allow a user to: 1) to copy the selected text, 2) generate a highlight stroke over the selected text; 3) generate a stroke to strike-through the text; and 4) generate a stroke to underline the text; and 5) toggle between linear and rectangle selection.
> 
>     * By double pressing or triple pressing with the linear selection tool, a user may select by words or lines, respectively, instead of by character/symbol.

AAAAAAAAAAAAAAAAAAAAAAAah, finally I will  perhaps throw away my wine/"PDF XChange Viewer" I use on linux since 10 years, because I never found a Linux program that highlight properly PDF files!

Thanks a lot

**Comment by Vozf** [comment]
copying on ctrl c would certainly be nice instead of pop up copy

**Comment by cycfunc** [comment]
Has this feature been pushed on to the latest version 1.1.3?

**Comment by jdohn15** [comment]
> Has this feature been pushed on to the latest version 1.1.3?

I'm trying to figure this out too... I cant select text within a pdf.

**Comment by Technius** [comment]
@cycfunc @jdohn15 PDF text selection is currently only available in the [nightly builds](https://github.com/xournalpp/xournalpp/releases/tag/nightly) and will be released in the upcoming version 1.2.0.
